libbeat/cmd/instance/locks/lock.go

```diff
@@ -30,233 +28,89 @@ import (
+	// as another beat can swoop in and lock the file before this beat removes it.
+	if runtime.GOOS != "windows" {
+		err := os.Remove(lock.fileLock.Path())
+		if err != nil {
+			lock.logger.Warnf("could not remove lockfile at %s", lock.fileLock.Path())
 		}
 	}
 
-	// if we're here, we've locked the file
-	// unlock so we can continue
-	err = lock.fileLock.Unlock()
+	err := lock.fileLock.Unlock()
 	if err != nil {
-		return fmt.Errorf("error unlocking a previously found file %s after a temporary lock", lock.filePath)
-	}
-
-	// read in whatever existing lockfile caused us to fail
-	pf, err := lock.readExistingPidfile()
-	// Case: two beats start up simultaneously, there's a chance we could "see" the pidfile before the other process writes to it
-	// or, the other beat died before it could write the pidfile.
-	// Sleep, read again. If we still don't have anything, assume the other PID is dead, continue.
-	if errors.Is(err, ErrLockfileEmpty) {
-		lock.logger.Debugf("Found other pidfile, but no data. Retrying.")
-		time.Sleep(time.Millisecond * 500)
-		pf, err = lock.readExistingPidfile()
-		if errors.Is(err, ErrLockfileEmpty) {
-			lock.logger.Debugf("No PID found in other lockfile, continuing")
-			return lock.recoverLockfile()
-		} else if err != nil {
-			return fmt.Errorf("error re-reading existing lockfile: %w", err)
-		}
-	} else if err != nil {
-		return fmt.Errorf("error reading existing lockfile: %w", err)
-	}
-
-	// Case: the lockfile is locked, but by us. Probably a coding error,
-	// and probably hard to do
-	if pf.PID == os.Getpid() {
-		// the lockfile was written before the beat started, meaning we restarted and somehow got the same pid
-		// in which case, continue
-		if lock.beatStart.Before(pf.WriteTime) {
-			return fmt.Errorf("lockfile for beat has been locked twice by the same PID, potential bug.")
-		}
-		lock.logger.Debugf("Beat has started with the same PID, continuing")
-		return lock.recoverLockfile()
+		return fmt.Errorf("unable to unlock data path: %w", err)
 	}
 
-	// Check to see if the PID found in the pidfile exists.
-	existsState, err := findMatchingPID(pf.PID)
-	// Case: we have a lockfile, but the pid from the pidfile no longer exists
-	// this was presumably due to the dirty shutdown.
-	// Try to reset the lockfile and continue.
-	if errors.Is(err, metricproc.ProcNotExist) {
-		lock.logger.Debugf("%s shut down without removing previous lockfile, continuing", lock.beatName)
-		return lock.recoverLockfile()
-	} else if err != nil {
-		return fmt.Errorf("error looking up status for pid %d: %w", pf.PID, err)
-	} else {
-		// Case: the PID exists, but it's attached to a zombie process
-		// In this case...we should be okay to restart?
-		if existsState == metricproc.Zombie {
-			lock.logger.Debugf("%s shut down without removing previous lockfile and is currently in a zombie state, continuing", lock.beatName)
-			return lock.recoverLockfile()
+	// now unlock on windows.
+	if runtime.GOOS == "windows" {
```

Comment:
could we split this into 3 files?  lock.go, unlock_posix.go, unlock_windows.go.  Then we can have the `Unlock` function in the OS specific file, and use build tags to only compile the "right" one.  The big benefit is that the doc strings can be specific to the OS since we are switching behavior based on that and it will be easier to understand later on.

Reply:
Ah, good idea.